### ISSUE TITLE
Reproducing Tables 1/4 with CUT3R: KITTI/7-Scenes Discrepancies and Reset/Inference Settings

### ISSUE THREAD

**Issue by SSSayon** [NVlabs/scal3r#2]
Hi, thanks for sharing Scal3R! I tried fine-tuning the CUT3R version using commit `af4c8e8`. I'm seeing something I haven't been able to explain: **KITTI ATE is quite a bit worse than in the paper, but the 7-Scenes reconstruction errors are actually lower.** The original CUT3R tracking results are close to the paper, so I'm wondering whether I've missed something in the Scal3R training or inference setup.

Here is what I get, compared with Tables 1 and 4 in the paper:

| Model | KITTI ATE ↓ (m) | 7-Scenes Acc. ↓ (m) | Comp. ↓ (m) | NC ↑ |
|---|---:|---:|---:|---:|
| CUT3R paper | 207.2 | 0.1300 | 0.0620 | 0.5440 |
| CUT3R mine | 208.02 | 0.1062 | 0.0503 | 0.5418 |
| Scal3R(CUT3R) paper | 69.7 | 0.0690 | 0.0350 | 0.5660 |
| Scal3R(CUT3R) mine | **107.45** | **0.0292** | **0.0199** | **0.5604** |

<details>
<summary>My setup and how I trained/evaluated</summary>

- **Environment:** Ubuntu 22.04, Python 3.11, PyTorch 2.5.1+cu124, CUDA 12.4, GTSAM 4.2; RTX 5880 Ada 48GB. Training used one GPU; independent evaluation runs used four GPUs.
- **Training:** `cut3r_512_dpt_4_64.pth`, frozen backbone, 7.11M trainable parameters (~0.89%); all original backbone tensors verified unchanged. Public TartanAir v1, 18 environments/Easy+Hard, 369 sequences/306,637 left RGB/depth frames, using `TartanAir_Multi` in place of the configured `TartanAirWds_Multi`. Four views, 4,000 samples/epoch, batch 8, AdamW/lr 1e-4, 40 epochs, BF16, seed 0; final checkpoint, without downstream checkpoint selection. ~9 h training.
- **Evaluation:** KITTI 00–10, full `image_2` sequences, Sim(3)-aligned ATE averaged over 11 sequences. Scal3R used the released per-sequence loop settings, K=12, non-KF buffer=8, reset=10, init=2, `--no_kf_gate`. For 7-Scenes: all 18 TestSplit sequences, first 300 consecutive frames, `--kf_every 1 --max_frames 300 --model_name scal3r`, default K=4/state gating; registered depth and the shared normalization/224×224 crop/ICP evaluator. CUT3R used its original branch.

</details>

<details>
<summary>Results of all 11 KITTI sequences</summary>

| Sequence | CUT3R paper | CUT3R mine | Scal3R paper | Scal3R mine |
|---|---:|---:|---:|---:|
| 00 | 191.00 | 190.32 | 45.30 | 117.25 |
| 01 | 644.10 | 639.16 | 165.00 | 153.00 |
| 02 | 295.40 | 302.35 | 139.90 | 278.85 |
| 03 | 150.10 | 153.05 | 33.90 | 44.05 |
| 04 | 20.10 | 22.20 | 9.90 | 22.43 |
| 05 | 155.60 | 152.51 | 29.40 | 93.90 |
| 06 | 132.70 | 133.03 | 47.40 | 77.71 |
| 07 | 72.80 | 73.26 | 6.40 | 63.56 |
| 08 | 231.80 | 231.38 | 227.00 | 218.97 |
| 09 | 206.20 | 209.38 | 29.60 | 42.43 |
| 10 | 179.60 | 181.57 | 33.20 | 69.82 |

</details>

---

Below, I looked more closely at **02, 05 and 07 because their ATE gaps were large**.

One thing I noticed is the timing of the pose-token initialization around a reset. In [`forward_recurrent`](https://github.com/NVlabs/scal3r/blob/af4c8e8047a4541ad573722b7d649221729a2d23/src/dust3r/model.py#L1325-L1340), the learned initial pose token is used on the view marked `reset=True`, but state/memory are reset **after** that view is processed. The boundary image is then processed again with `reset=False`.

**This seems to create a one-step mismatch: the initial pose token is used on the last step with the old state, while the first step with the freshly reset state/memory instead retrieves its pose token from memory.** If the initial token is intended to accompany a fresh recurrent state, this could introduce an inconsistent pose initialization at each reset boundary.

I tried moving that token initialization to the first view **after** the reset:

```python
# Released code
if i == 0 or reset_mask:
# What I tried
if i == 0 or (i > 0 and bool(views[i - 1]["reset"].any())):
```

| KITTI sequence | Before | After this change | Paper (Table 7) |
|---|---:|---:|---:|
| 02 | 278.85 | 281.19† | 139.88 |
| 05 | 93.90 | **24.21** | 29.37 |
| 07 | 63.56 | **9.29** | 6.44 |

† On Seq. 02, iSAM2 fails and disables itself after the first loop closure.

This helps ATE on 05/07, but I don't think it's enough to call it a fix: on 05, rotation/translation RPE actually get worse, from 1.30°/0.278 m to 1.74°/0.904 m. I also haven't rerun the full benchmark with this change.

Below is the CUT3R trajectory digitized from Fig. 7 alongside my two runs on 05, on the same axes. My trajectories use 3D Sim(3) alignment to ground truth.

<img width="2240" height="832" alt="Image" src="https://github.com/user-attachments/assets/07661a73-1194-46d1-b968-9dd971d44b6c" />

I wonder if you could clarify a few things:

1. Is the pose-token/reset timing above intentional? And what would you recommend for the iSAM2 failure on 02?
2. Which inference settings should I use to match Tables 1/4? The KITTI script uses init=2, `--no_kf_gate` and 10-frame resets, while the paper describes init=5/KF-gated updates and mentions both frames and keyframes for the reset interval. Also, `eval/mv_recon/run.sh` uses `model_name='ours'` (which takes the vanilla CUT3R branch) and doesn't set the documented 300-frame flags. Is `--model_name scal3r --kf_every 1 --max_frames 300` the right choice?
3. What might explain the better 7-Scenes reconstruction results in my reproduction?

Thanks again for your great work and for any guidance you can provide!
